crates/ibc/src/core/ics02_client/context.rs

```diff
@@ -1,9 +1,33 @@
-use crate::core::{
-    ics24_host::path::{ClientConsensusStatePath, ClientStatePath},
-    ContextError,
-};
+use crate::core::ics02_client::client_state::ClientState;
+use crate::core::ics02_client::consensus_state::ConsensusState;
+use crate::core::ics24_host::identifier::ClientId;
+use crate::core::ics24_host::path::ClientConsensusStatePath;
+use crate::core::ics24_host::path::ClientStatePath;
+use crate::core::ContextError;
 
-use super::{client_state::ClientState, consensus_state::ConsensusState};
+/// Defines the types that all clients must implement.
+pub trait ClientTypes {
+    type V: ClientValidationContext;
+    type E: ClientExecutionContext;
+    type AnyClientState: ClientState<Self::V, Self::E>;
+    type AnyConsensusState: ConsensusState;
+}
+
+/// Defines the methods that all client `ValidationContext`s (precisely the
+/// generic parameter of
+/// [`crate::core::ics02_client::client_state::ClientStateValidation`] ) must
+/// implement.
+pub trait ClientValidationContext: ClientTypes + Sized {
```

Comment:
> With the ClientTypes supertrait, we claim that every client's ValidationContext will need:
access to consensus_state() and client_state() methods
types AnyClientState, AnyConsensusState, ClientValidationContext and ClientExecutionContext
We can already see that the Tendermint client doesn't need the client_state() method, nor the type AnyClientState, V and E.

Tendermint itself doesn't use it, but isn’t that much tailored specifically for the Tendermint clients?
I'm referring to the implicit design choice we made. We've granted the mentioned types to the execution trait, meaning clients have the write access. Based on this premise, shouldn't clients also have the read access? be aware of their own states? If someone wants to use it for validation, they can, even though it might not be utilized by all, like in the case of Tendermint.

Plus, my view was a bit beyond the requirements of light clients alone. The `ClientValidationContext` interface provides added flexibility in terms of granting client validation access to any components if needed. This can include a light client or even an application (e.g. in case of `SendPacketValidationContext`)

Reply:
> Tendermint itself doesn't use it, but isn’t that much tailored specifically for the Tendermint clients?

It's the other way around: the Tendermint client works as a counter example here. Basically, if I can show you any client where its `ValidationContext` doesn't need one of the methods you're forcing it to have with the `ClientValidationContext` trait, then this shows a weakness in the design.

I think the `ClientExecutionContext` confused you here. Only until the very end did I add the `ClientExecutionContext`; the whole thing works without it. I only found that since our API explicitly gives the responsibility to clients to store their client & consensus states, then perhaps it would be clearer for everyone if we had a trait (`ClientExecutionContext`) which gave a uniform interface for all the clients to use to perform these 2 tasks. Note that we don't *need* it; every client *could* define their own `store_client_state()` and `store_consensus_state()`. Except now, that would make it quite tedious for the host, as each would have a slightly different interface, but semantically would achieve the same outcome (i.e. storing a client state and consensus state).

> Based on this premise, shouldn't clients also have the read access? be aware of their own states?

They absolutely can. Every client defines its own `ValidationContext`; if they need to read the client or consensus state, then all they need to do is add a method `client_state()` or `consensus_state()`. My point is: we don't know explicitly know if that will be necessary, so why enforce it with a `ClientValidation` trait bound? In other words, adding a `ClientValidation` trait bound is *limiting* because it enforces a structure on their contexts that we don't know if they will need, whereas not putting any trait bound lets them define their context however they see fit (including but not limited to reading their own `ClientState` and `ConsensusState`).

> Plus, my view was a bit beyond the requirements of light clients alone. The ClientValidationContext interface provides added flexibility in terms of granting client validation access to any components if needed. This can include a light client or even an application (e.g. in case of SendPacketValidationContext)

I don't follow here. `send_packet()` is part of `core` (i.e. we control the implementation), so there's no uncertainty as to which `Context` method will be needed.

Similarly for applications (i.e. implementing the `Module` trait), if we were to start using this pattern, imposing `ClientValidation` as a trait bound is *more limiting* than not putting any trait bound at all.

Reply:
> (1) Depending on how hosts define context structs, a design can originate from a wrong impl, which imho necessitated a revision of MockContext to get it to the correct definition.

What would be a "correct" `MockContext`? IMO it's fine as is, exactly because our design should not prevent the user from implementing all the traits on the same struct - but neither should it *force* the user to. AFAIK, the current ADR7 design achieves this goal.

> (2) Given your explanation, ADR07 ideally shouldn't function with the current mock setup.

Why is that?